packages/rum-core/src/domain/contexts/pageStateContexts.ts

```diff
@@ -0,0 +1,122 @@
+import type { Duration, RelativeTime } from '@datadog/browser-core'
+import {
+  isExperimentalFeatureEnabled,
+  noop,
+  addDuration,
+  addEventListeners,
+  relativeNow,
+  DOM_EVENT,
+} from '@datadog/browser-core'
+
+export const PAGE_STATE_CONTEXT_MAX_LENGTH = 500
+
+export const enum PageState {
+  ACTIVE = 'active',
+  PASSIVE = 'passive',
+  HIDDEN = 'hidden',
+  FROZEN = 'frozen',
+  TERMINATED = 'terminated',
+}
+
+export type PageStateContext = Array<{ state: PageState; startTime: RelativeTime }>
+
+export interface PageStateContexts {
+  getPageStates: (startTime: RelativeTime, duration: Duration) => PageStateContext | undefined
+  stop: () => void
+}
+
+let pageStateContext: PageStateContext = []
+let state: PageState | undefined
+
+export function startPageStateContexts(): PageStateContexts {
+  if (!isExperimentalFeatureEnabled('resource_page_states')) {
+    return {
+      getPageStates: () => undefined,
+      stop: noop,
+    }
+  }
+
+  addPageState(getState())
+
+  const { stop } = addEventListeners(
+    window,
+    [
+      DOM_EVENT.PAGE_SHOW,
+      DOM_EVENT.FOCUS,
+      DOM_EVENT.BLUR,
+      DOM_EVENT.VISIBILITY_CHANGE,
+      DOM_EVENT.RESUME,
+      DOM_EVENT.FREEZE,
+      DOM_EVENT.PAGE_HIDE,
+    ],
+    (event) => {
+      // Only get events fired by the browser to avoid false state changes done with custom events
+      // cf: developer extension auto flush: https://github.com/DataDog/browser-sdk/blob/2f72bf05a672794c9e33965351964382a94c72ba/developer-extension/src/panel/flushEvents.ts#L11-L12
+      if (!event.isTrusted) {
+        return
+      }
+
+      if (event.type === DOM_EVENT.FREEZE) {
+        addPageState(PageState.FROZEN)
+      } else if (event.type === DOM_EVENT.PAGE_HIDE) {
+        addPageState((event as PageTransitionEvent).persisted ? PageState.FROZEN : PageState.TERMINATED)
+      } else {
+        addPageState(getState())
+      }
+    },
+    { capture: true }
+  )
+
+  return {
+    getPageStates(startTime: RelativeTime, duration: Duration) {
+      const context: PageStateContext = []
+      const endTime = addDuration(startTime, duration)
+      for (let i = pageStateContext.length - 1; i >= 0; i--) {
+        const { startTime: stateStartTime } = pageStateContext[i]
+
+        if (stateStartTime >= endTime) {
+          continue
+        }
+
+        context.unshift(pageStateContext[i])
+
+        if (stateStartTime < startTime) {
+          break
+        }
+      }
+
+      return context.length ? context : undefined
+    },
+    stop,
+  }
+}
+
+export function getState(): PageState {
+  if (document.visibilityState === 'hidden') {
+    return PageState.HIDDEN
+  }
+  if (document.hasFocus()) {
+    return PageState.ACTIVE
+  }
+  return PageState.PASSIVE
+}
+
+export function addPageState(nextState: PageState, pageStateContextMaxLength = PAGE_STATE_CONTEXT_MAX_LENGTH) {
```

Comment:
<!-- Thoughts represent an idea that popped up from reviewing. These comments are non-blocking by nature, but they are extremely valuable and can lead to more focused initiatives and mentoring opportunities.  -->
**💭 thought:** ‏overriding the limit from here seems odd, what about providing this limit in the start method instead?

Reply:
While it may be a bit odd, do you think it's good enough for now? According to your [comment](https://github.com/DataDog/browser-sdk/pull/1890#discussion_r1062589880) if this code has to leave longer than the investigation it will be refactored. So we could wait that moment to refine tests.

Reply:
Sure, worth to keep it in mind though.